src/action/file_stat.rs

```diff
@@ -0,0 +1,532 @@
+        .map_or(None, |dur| Some(dur.as_secs())));
+}
+
+#[cfg(target_os = "linux")]
+fn get_status_change_time(metadata: &Metadata) -> Option<SystemTime> {
+    use std::time::Duration;
+    use std::os::unix::fs::MetadataExt;
+
+    UNIX_EPOCH.checked_add(Duration::from_secs(metadata.ctime() as u64))
+}
+
+#[cfg(target_os = "linux")]
+fn form_response(original_path: &PathBuf, destination: &PathBuf)
+                 -> Result<Response, std::io::Error> {
+    use std::os::unix::fs::MetadataExt;
+
+    let metadata = fs::symlink_metadata(destination)?;
+    let original_metadata = fs::symlink_metadata(original_path)?;
+
+    Ok(Response {
+        mode: Some(metadata.mode() as u64),
+        inode: Some(metadata.ino() as u32),
+        device: Some(metadata.dev() as u32),
+        hard_links: Some(metadata.nlink() as u32),
+        uid: Some(metadata.uid() as u32),
+        gid: Some(metadata.gid() as u32),
+        size: Some(metadata.size() as u64),
+        access_time: get_time_option(metadata.accessed()),
+        modification_time: get_time_option(metadata.modified()),
+        status_change_time: get_status_change_time(&metadata),
+        blocks_number: Some(metadata.blocks() as u32),
+        block_size: Some(metadata.blksize() as u32),
+        represented_device: Some(metadata.rdev() as u32),
+        flags_linux: get_linux_flags(destination),
+
+        symlink: match original_metadata.file_type().is_symlink() {
+            true => Some(fs::read_link(original_path).unwrap()),
+            false => None
+        },
+
+        pathspec: PathSpec {
+            nested_path: None,
+            path: Some(original_path.clone()),
+        },
+
+        extended_attributes: vec![],
+    })
+}
+
+#[cfg(not(target_os = "linux"))]
+fn form_response(original_path: &PathBuf, destination: &PathBuf)
+                 -> Result<Response, std::io::Error> {
+    Ok(Response::default())
+}
+
+fn collapse_pathspec(pathspec: PathSpec) -> PathBuf {
+    fn recursive_collapse(pathspec: PathSpec) -> PathBuf {
+        match pathspec.path {
+            Some(path) => {
+                match pathspec.nested_path {
+                    Some(nested_path_box) => path.join(recursive_collapse(*nested_path_box)),
+                    None => path,
+                }
+            }
+            None => PathBuf::default()
+        }
+    }
+
+    let mut result = recursive_collapse(pathspec);
+    if !result.has_root() {
+        result = PathBuf::from("/").join(result);
+    }
+    result
+}
+
+fn get_path(path: &Option<String>) -> Option<PathBuf> {
+    match path {
+        Some(string_path) => Some(PathBuf::from(string_path)),
+        _ => None,
+    }
+}
+
+#[cfg(target_os = "linux")]
+fn get_linux_flags(path: &Path) -> Option<u32> {
+    use std::os::raw::c_long;
+    use std::os::unix::io::AsRawFd;
+
+    let file = match File::open(path) {
+        Ok(file) => file,
+        Err(_) => return None,
```

Comment:
I don't think we should ignore such errors, it warrants a log entry.

Reply:
Fixed.